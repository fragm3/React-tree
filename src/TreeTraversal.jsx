import React, { Component } from 'react';
import './App.css';
import Menu from './Menu';
import { Transition } from 'react-move';

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinaryTree {
    constructor(root) {
        this.root = root;
        this.list = [];
    }

    // returns root of the tree
    getRootNode = () => {
        return this.root;
    }

    inorder = (node) => {
        if(node !== null) {
            this.inorder(node.left);
            this.list.push(node.data);
            this.inorder(node.right);
        }
    }

    preorder = (node) => {
        if(node != null) {
            this.list.push(node.data);
            this.preorder(node.left);
            
            this.preorder(node.right);
        }
    }

    postorder = (node) => {
        if(node != null) {
            this.postorder(node.left);
            this.postorder(node.right);
            this.list.push(node.data);
        }
    }

    resetList = () => {
        this.list = [];
    }

    getList = () => {
        return this.list;
    }

    bft = (node) => {
        if(node != null) {
            var arr = [];
            arr.push(node);
            while(arr.length != 0){
                var element = arr.shift();
                this.list.push(element.data);
                console.log(element.data);
                if(element.left != null)
                    arr.push(element.left);
                if(element.right != null)
                    arr.push(element.right);    
            }
        }
    }
}

var node = new Node("34");
node.left = new Node("23");
node.left.left = new Node("12");
node.left.right = new Node("04");
node.left.right.left = new Node("16");
node.left.right.right = new Node("09");
node.right = new Node("92");
var BT = new BinaryTree(node);

var root = BT.getRootNode();


export default class TreeTraversal extends Component{
	constructor(){
        super();
        this.state = {
            list: [],
            isPlaying: false
        };
        
    }

    array = [];

    // prints 34 23 92 12 04 16 09
    bft = ()  => {
        BT.resetList();
        this.setState({list: []})
        BT.bft(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 34 23 12 04 16 09 92
    preOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.preorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 12 23 16 04 09 34 92
    inOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.inorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 12 16 09 04 23 92 34
    postOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.postorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }

    reset = () => {
        this.setState({isPlaying: false, list: []});
    }
	
	render() {
        let {list} = this.state;
        let renderElement = null;
        if(this.state.list){
        renderElement = (
            <div>
                <button onClick={this.reset}>Reset</button>
                <Transition
                   data={list}
                   getKey={d => d}
                   update={d => ({
                   translate: 0,
                   opacity: 1
                   })}
                   enter={d => ({
                       translate: 1,
                       opacity: 0
                   })}
                   leave={d => ({
                       translate: -1,
                       opacity: 0
                   })}
                   stagger={200}
                >
                {dataArray => (
                    <div className = "traverse-number-container">
                    {dataArray.map(data => (
                    <li
                      className="traverse-number"
                      key={data.key}
                      style={{opacity: data.state.opacity}}
                    >
                      {data.key}
                    </li>
                    ))
                    }
                    </div>
                )}
            </Transition>
        </div>
        )
    }

		return (
            <div className = "tree-traversal">
                <Menu
                  bft={this.bft} 
                  preorder={this.preOrder}
                  inorder={this.inOrder}
                  postOrder={this.postOrder}
                />
                {renderElement}
            </div>    
        )
	}
};