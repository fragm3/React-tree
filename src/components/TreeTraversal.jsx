import React, { Component } from 'react';
import '../App.css';
import Menu from './Menu';
import { Transition } from 'react-move';
import {Node, BinaryTree} from './Tree'

var node = new Node("34");
node.left = new Node("23");
node.left.left = new Node("12");
node.left.right = new Node("04");
node.left.right.left = new Node("16");
node.left.right.right = new Node("09");
node.right = new Node("92");
var BT = new BinaryTree(node);

var root = BT.getRootNode();

export default class TreeTraversal extends Component{
	constructor(){
        super();
        this.state = {
            list: [],
            isPlaying: false
        };
        
    }

    // prints 34 23 92 12 04 16 09
    bft = ()  => {
        BT.resetList();
        this.setState({list: []})
        BT.bft(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 34 23 12 04 16 09 92
    preOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.preorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 12 23 16 04 09 34 92
    inOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.inorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }
    
    // prints 12 16 09 04 23 92 34
    postOrder = () => {
        BT.resetList();
        this.setState({list: []})
        BT.postorder(root);
        var list = BT.getList();
        this.setState({list: list})
    }

    reset = () => {
        this.setState({isPlaying: false, list: []});
    }
	
	render() {
        let {list} = this.state;
        let renderElement = null;
        if(this.state.list){
        renderElement = (
            <div style={{marginTop: '20px'}}>
                <button onClick={this.reset}>Reset</button>
                <Transition
                    data={list}
                    getKey={d => d}
                    update={d => ({
                    translate: 0,
                    opacity: 1
                    })}
                    enter={d => ({
                        translate: 10,
                        opacity: 0
                    })}
                    leave={d => ({
                        translate: -1,
                        opacity: 0
                    })}
                    stagger={400}
                >
                {dataArray => (
                    <div className = "traverse-number-container">
                    {dataArray.map(data => (
                    <li
                        className="traverse-number"
                        key={data.key}
                        style={{
                        opacity: data.state.opacity,
                        transform: `translate(${10 * data.state.translate}px)`
                        }}
                    >
                        {data.key}
                    </li>
                    ))
                    }
                    </div>
                )}
            </Transition>
        </div>
        )
    }

		return (
            <div className = "tree-traversal">
                <Menu
                    bft={this.bft} 
                    preorder={this.preOrder}
                    inorder={this.inOrder}
                    postOrder={this.postOrder}
                />
                {renderElement}
            </div>    
        )
	}
};