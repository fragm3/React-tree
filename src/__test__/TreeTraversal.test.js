import React from 'react';
import { shallow } from 'enzyme';
import TreeTraversal from '../components/TreeTraversal.jsx';
import renderer from 'react-test-renderer'

describe('<TreeTraversal />', () => {
  const wrapper = shallow(<TreeTraversal />);
  it('renders TreeTraversal without crashing', () => {
    wrapper
  })
  it('TreeTraversal matches snapshot', () => {
    const tree = renderer.create(<TreeTraversal/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('playing is set false', ()=> {
    expect(wrapper.state('isPlaying')).toEqual(false);
  })
  it('BFS Traversal start', ()=> {
    let arr = ["34","23","92","12","04","16","09"];
    wrapper.instance().bft();
    expect(wrapper.state('list')).toEqual(arr);
  })
  it('InOrder Traversal start', ()=> {
    let arr = ["12","23","16","04","09","34","92"];
    wrapper.instance().inOrder();
    expect(wrapper.state('list')).toEqual(arr);
  })
  it('PreOrder Traversal start', ()=> {
    let arr = ["34", "23", "12", "04", "16", "09", "92"];
    wrapper.instance().preOrder();
    expect(wrapper.state('list')).toEqual(arr);
  })
  it('postOrder Traversal start', ()=> {
    let arr = ["12","16","09","04","23","92","34"];
    wrapper.instance().postOrder();
    expect(wrapper.state('list')).toEqual(arr);
  })
});
