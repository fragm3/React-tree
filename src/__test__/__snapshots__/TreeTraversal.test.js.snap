// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeTraversal /> TreeTraversal matches snapshot 1`] = `
<div
  className="tree-traversal"
>
  <div
    className="menu-container"
  >
    <div>
      <button
        className="menu-button"
        onClick={[Function]}
      >
        Select traversal
        <span
          className="caret"
        />
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "marginTop": "20px",
      }
    }
  >
    <button
      onClick={[Function]}
    >
      Reset
    </button>
    <div
      className="traverse-number-container"
    />
  </div>
</div>
`;
